internal/layers/wcow_parse.go

```diff
@@ -0,0 +1,183 @@
+	if err != nil {
+		return nil, err
+	}
+	parentCimPaths, err := getParentCimPaths(m)
+	if err != nil {
+		return nil, err
+	}
+	if len(parentLayerPaths) != len(parentCimPaths) {
+		return nil, fmt.Errorf("invalid mount, number of parent layer paths & cim paths should be same")
+	}
+	forkedCimLayers := []forkedCIMLayer{}
+	for i := 0; i < len(parentCimPaths); i++ {
+		forkedCimLayers = append(forkedCimLayers, forkedCIMLayer{
+			layerPath: parentLayerPaths[i],
+			cimPath:   parentCimPaths[i],
+		})
+	}
+	return &wcowForkedCIMLayers{
+		scratchLayerData: scratchLayerData{
+			scratchLayerPath: m.Source,
+		},
+		layers: forkedCimLayers,
+	}, nil
+}
+
+// ParseWCOWLayers parses the layers provided by containerd into the format understood by hcsshim and prepares
+// them for mounting.
+func ParseWCOWLayers(rootfs []*types.Mount, layerFolders []string) (WCOWLayers, error) {
+	if err := validateRootfsAndLayers(rootfs, layerFolders); err != nil {
+		return nil, err
+	}
+
+	if len(layerFolders) > 0 {
+		return &wcowWCIFSLayers{
+			scratchLayerData: scratchLayerData{
+				scratchLayerPath: layerFolders[len(layerFolders)-1],
+			},
+			layerPaths: layerFolders[:len(layerFolders)-1],
+		}, nil
+	}
+
+	m := rootfs[0]
+	switch m.Type {
+	case LegacyMountType:
+		parentLayers, err := getParentPaths(m)
+		if err != nil {
+			return nil, err
+		}
+		return &wcowWCIFSLayers{
+			scratchLayerData: scratchLayerData{
+				scratchLayerPath: m.Source,
+			},
+			layerPaths: parentLayers,
+		}, nil
+	case CimFSMountType:
+		return parseForkedCimMount(m)
+	default:
+		return nil, fmt.Errorf("invalid windows mount type: '%s'", m.Type)
+	}
+}
+
+// Parse the layer data to get the UVM boot files from the layers.
+func ParseWCOWUVMBootFilesFromLayers(ctx context.Context, rootfs []*types.Mount, layerFolders []string) (*uvm.WCOWBootFiles, error) {
+	var parentLayers []string
+	var scratchLayer string
+	var err error
+
+	if err = validateRootfsAndLayers(rootfs, layerFolders); err != nil {
+		return nil, err
+	}
+
+	if len(layerFolders) > 0 {
+		parentLayers = layerFolders[:len(layerFolders)-1]
+		scratchLayer = layerFolders[len(layerFolders)-1]
+	} else {
+		m := rootfs[0]
+		switch m.Type {
+		case LegacyMountType:
+			parentLayers, err = getParentPaths(m)
+			if err != nil {
+				return nil, err
+			}
+			scratchLayer = m.Source
+		default:
+			return nil, fmt.Errorf("mount type '%s' is not supported for UVM boot", m.Type)
+		}
+	}
+
+	uvmFolder, err := uvmfolder.LocateUVMFolder(ctx, parentLayers)
+	if err != nil {
+		return nil, fmt.Errorf("failed to locate utility VM folder from layer folders: %w", err)
+	}
+
+	// In order for the UVM sandbox.vhdx not to collide with the actual
+	// nested Argon sandbox.vhdx we append the \vm folder to the last
+	// entry in the list.
+	scratchLayer = filepath.Join(scratchLayer, "vm")
+	scratchVHDPath := filepath.Join(scratchLayer, "sandbox.vhdx")
+	if err = os.MkdirAll(scratchLayer, 0777); err != nil {
```

Comment:
I agree. However, this is just for backwards compatibility so we can remove this code in the future. Plus, if we don't create these files here, we will have to pass all the layer paths to the UVM creator function (via the `WCOWBootFiles` struct) which I am trying to avoid. What do you think? Let's ask @kevpar  too.

Reply:
I agree having this in a function named `Parse` is non-ideal, but it is work we want to get out of the way initially so we don't have to worry about it later. Maybe there is a better name for this function which conveys that it does some setup work too?